### ISSUE TITLE
Use keystore as default secure storage for Android

### ISSUE THREAD

**Issue by mCodex** [mCodex/react-native-sensitive-info#196]
Since v6.0.0 the lib should use Android's Keystore as default storage for better security purposes and shared preferences could be still used, but not as default.

**Comment by kneza23** [comment]
Official RN docs included your library to the Security page https://reactnative.dev/docs/security. It would be amazing if eventually we get Android's Keystore as default storage :)

**Comment by mCodex** [comment]
@kneza23 thanks for pointing it out! I didn't know 😱  Yeah, I think it is time to do this migration

**Comment by mCodex** [comment]
This afternoon I will create a new branch from master and I'll merge keystore into that branch. So, we gonna have a starting point for this 👏 👏 👏

**Comment by mCodex** [comment]
Already merged in [keystoreAsDefault](/github/linguist/tree/feature/keystoreAsDefault). Also, it includes the  new example app, but it is a WIP 😄

**Comment by mCodex** [comment]
There are a new release available in npm which includes keystore by default. For more information, please read [here](https://github.com/mCodex/react-native-sensitive-info/releases/tag/v6.0.0).

**Comment by juanapp** [comment]
Is there a solution for data migration? I would love to integrate this version but I'm also afraid that it will cause quite some troubles to users that have data saved..

**Comment by mCodex** [comment]
> Is there a solution for data migration? I would love to integrate this version but I'm also afraid that it will cause quite some troubles to users that have data saved..

Unfortunately there is not. Maybe we need an extra method to read from sharedPreferences without using keystore something like `getAllUncryptedDataFromSharedPreferences` and then every users should handle saving all these infos again using `setItem` and now everything will be encrypted in Android.

**Comment by juanapp** [comment]
@mCodex That would be ideal... I think what we need is to have a way to trigger the migration from everything that is stored on SharedPreferences to the KeyStore, keeping the same structure and then deleting it from SharedPreferences. That would save the hassle of the first time run after upgrading.

**Comment by mCodex** [comment]
> @mCodex That would be ideal... I think what we need is to have a way to trigger the migration from everything that is stored on SharedPreferences to the KeyStore, keeping the same structure and then deleting it from SharedPreferences. That would save the hassle of the first time run after upgrading.

Yeah, I couldn't agree more. This is why v6.0 is yet in alpha release. I don't want to developers updating RNInfo and their users losing data. On the other hand, there is a work to do to handle this scenario but no one is taking care of it right now. I'll see if I can code something to speed up this release.

**Comment by kgsachinthaudara** [comment]
@mCodex  when are your hope to release v6, we are waiting for great news ❤️

**Comment by idanlevi1** [comment]
Can you write the changes between alpha versions? 
and the alpha version is stable to use in production new?

**Comment by Marvedog** [comment]
> > @mCodex That would be ideal... I think what we need is to have a way to trigger the migration from everything that is stored on SharedPreferences to the KeyStore, keeping the same structure and then deleting it from SharedPreferences. That would save the hassle of the first time run after upgrading.
> 
> Yeah, I couldn't agree more. This is why v6.0 is yet in alpha release. I don't want to developers updating RNInfo and their users losing data. On the other hand, there is a work to do to handle this scenario but no one is taking care of it right now. I'll see if I can code something to speed up this release.

Do you need any help with this?

**Comment by mCodex** [comment]
> > > @mCodex That would be ideal... I think what we need is to have a way to trigger the migration from everything that is stored on SharedPreferences to the KeyStore, keeping the same structure and then deleting it from SharedPreferences. That would save the hassle of the first time run after upgrading.
> > 
> > 
> > Yeah, I couldn't agree more. This is why v6.0 is yet in alpha release. I don't want to developers updating RNInfo and their users losing data. On the other hand, there is a work to do to handle this scenario but no one is taking care of it right now. I'll see if I can code something to speed up this release.
> 
> Do you need any help with this?
> 
> @mCodex

Hi Marcus, of course! Being honest I still haven't looked into this yet.

PRs are always welcome.

**Comment by Marvedog** [comment]
@mCodex I'll take a look then

**Comment by aarondail** [comment]
In lieu of some kinda wholesale migration to deal with old unencrypted values in shared preferences, we have cooked up something super simple for our app: when we call `getItem()` we wrap it in a `try/catch` and (if it fails) then call `getItem()` again with a `{ dontDecrypt: true }` option we patched into the library.

The patch is just a one line change to `RNSensitiveInfoModule`'s `getItem()` method:
```
    @ReactMethod
    public void getItem(String key, ReadableMap options, Promise pm) {
        // ...omitted...
        if (value != null && options.hasKey("touchID") && options.getBoolean("touchID")) {
            // ...omitted...
//****
// The following line is the only one we changed (it used to be just else if (value !== null) {)
        } else if (value != null && !(options.hasKey("dontDecrypt") && options.getBoolean("dontDecrypt"))) {
//****
            try {
                pm.resolve(decrypt(value));
            } catch (Exception e) {
                pm.reject(e);
            }
        } else {
            pm.resolve(value);
        }
    }
```

Again: super simple. Maybe too simple... and it does require the caller to do some work so still a breaking change. This could be made completely transparent (to callers) by dropping the `dontDecrypt` option we added and instead changing `pm.reject(e);` to `pm.resolve(value)` in the function I pasted above... though that may be dissatisfying too. Definitely not as satisfying as a real wholesale migration. But if this or some variant of it seems like something you'd like to have in the library I am happy to submit a PR.

**Comment by kneza23** [comment]
Hi guys. Any news or estimates on this?

**Comment by kam89** [comment]
Hi, is it possible to include fail attempt for each biometric authentication for both Android and iOS?

**Comment by fedeerbes** [comment]
Hi, is it possible to include #264 as well? That will greatly improve the user experience.

**Comment by tero-paananen** [comment]
I am using  `"react-native-sensitive-info": "^6.0.0-alpha.9`" on React Native Windows `0.64.7`. I think that module autolinking does not work, it is not done after calling `npx react-native autolink-windows`. There is no any changes into `AutolinkedNativeModules.g.cs`

**Comment by tero-paananen** [comment]
> I am using `"react-native-sensitive-info": "^6.0.0-alpha.9`" on React Native Windows `0.64.7`. I think that module autolinking does not work, it is not done after calling `npx react-native autolink-windows`. There is no any changes into `AutolinkedNativeModules.g.cs`

Manual linking seems to work in App.xaml.cs `PackageProviders.Add(new RNSensitiveInfoCPP.ReactPackageProvider());`

**Comment by MaxChenMC** [comment]
Hi. May I know whether 6.0.0 supports Face Recognition for Android?

**Comment by JeffreyLeeDave** [comment]
I would like to propose a delete all function

**Comment by ShivamJoker** [comment]
Is macOS support added ?

**Comment by davux** [comment]
Are there any news on the data migration part?

Version 6.x is documented as stable but it looks like it's actually still considered alpha because of the data migration. So I think many projects are stuck with 5.x until there is either a silent migration or at least some workaround to ensure no data is lost.

**Comment by mikeschoneveld** [comment]
> Are there any news on the data migration part?
> 
> Version 6.x is documented as stable but it looks like it's actually still considered alpha because of the data migration. So I think many projects are stuck with 5.x until there is either a silent migration or at least some workaround to ensure no data is lost.

Any update on the issue? The data migration is a very important part for us because we don't want all our users to login again.

**Comment by ender-null** [comment]
Is there a way to opt out of encryption? one of the things that I value a lot is backing up the data, and in the past I used a library on flutter that allowed for optional encryption on android, but all the libraries I saw on react-native use encryption or don't use SharedPreference.

**Comment by mCodex** [comment]
After five years since the last major leap, we’re shipping react-native-sensitive-info v6.0.0-rc.2, a Nitro-powered rearchitecture that keeps secure storage on the metal with Swift/Kotlin hybrids instead of the legacy bridge.

Yesterday’s v5.6.0 back-ports critical fixes and docs for teams that must stay on the classic bridge, giving everyone a stable landing zone before the RC era begins.

We modernized the internals without losing Android 9 credential fallbacks, StrongBox upgrades, or biometric parity.

The new RC bundles revamped TypeScript types, Android/iOS capability detection, and the refreshed light-theme playground showcasing automatic Class 3 biometric enforcement.

Migration is straightforward: existing JS calls keep working, but you gain Nitro speed, richer metadata, and stricter hardware guards by default.

Want to compare code paths? The entire v5 line remains available at v5 branch while the RC lives on master for early adopters.
